analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Verifier.cs

```diff
@@ -0,0 +1,106 @@
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.UnitTest.Helpers;
+
+namespace SonarAnalyzer.UnitTest.TestFramework
+{
+    internal class Verifier
+    {
+        private static readonly Regex ImportsRegexVB = new(@"^\s*Imports\s+.+$", RegexOptions.Multiline | RegexOptions.RightToLeft);
+        private readonly VerifierBuilder builder;
+        private readonly DiagnosticAnalyzer[] analyzers;
+        private readonly AnalyzerLanguage language;
+
+        public Verifier(VerifierBuilder builder)
+        {
+            this.builder = builder ?? throw new ArgumentNullException(nameof(builder));
+            analyzers = builder.Analyzers.Select(x => x()).ToArray();
+            if (!analyzers.Any())
+            {
+                throw new ArgumentException(nameof(builder.Analyzers) + " cannot be empty.");
+            }
+            if (analyzers.Any(x => x == null))
+            {
+                throw new ArgumentException(nameof(builder.Analyzers) + " cannot produce null.");
+            }
+            var allLanguages = analyzers.SelectMany(x => x.GetType().GetCustomAttributes<DiagnosticAnalyzerAttribute>()).SelectMany(x => x.Languages).Distinct().ToArray();
+            if (allLanguages.Length > 1)
+            {
+                throw new ArgumentException(nameof(builder.Analyzers) + " cannot declare different languages in DiagnosticAnalyzerAttribute.");
+            }
+            language = AnalyzerLanguage.FromName(allLanguages.Single());
+            if (!builder.Paths.Any())
+            {
+                throw new ArgumentException(nameof(builder.Paths) + " cannot be empty.");
+            }
+            if (builder.Paths.FirstOrDefault(x => !Path.GetExtension(x).Equals(language.FileExtension, StringComparison.OrdinalIgnoreCase)) is { } unexpectedPath)
+            {
+                throw new ArgumentException($"Path '{unexpectedPath}' doesn't match {language.LanguageName} file extension '{language.FileExtension}'.");
+            }
+        }
+
+        public void Verify()    // This should never has any arguments
+        {
+            using var scope = new EnvironmentVariableScope { EnableConcurrentAnalysis = true };         // ToDo: Implement properly
+            var paths = builder.Paths.Select(x => Path.GetFullPath(Path.Combine("TestCases", x)));
+            var pathsWithConcurrencyTests = paths.Count() == 1 ? CreateConcurrencyTest(paths) : paths;  // ToDo: Redesign when implementing concurrency
+            var solution = SolutionBuilder.CreateSolutionFromPaths(pathsWithConcurrencyTests, OutputKind.DynamicallyLinkedLibrary, builder.References);
+            foreach (var compilation in solution.Compile(builder.ParseOptions.ToArray()))
+            {
+                DiagnosticVerifier.Verify(compilation, analyzers, CompilationErrorBehavior.Default, null, null);
+            }
+        }
+
+        private IEnumerable<string> CreateConcurrencyTest(IEnumerable<string> paths)
```

Comment:
This is present also in the OldVerifier (following method too). The plan is to remove it later?

Reply:
Yes, all private methods except `RemoveTestCasesPrefix` should go away from `OldVerifier`.